provider/kubernetes/kubernetes.go

```diff
@@ -454,18 +458,23 @@ func getTLS(ingress *extensionsv1beta1.Ingress, k8sClient Client) ([]*tls.Config
 	return tlsConfigs, nil
 }
 
-func endpointPortNumber(servicePort corev1.ServicePort, endpointPorts []corev1.EndpointPort) int {
-	if len(endpointPorts) > 0 {
-		// name is optional if there is only one port
-		port := endpointPorts[0]
-		for _, endpointPort := range endpointPorts {
-			if servicePort.Name == endpointPort.Name {
-				port = endpointPort
-			}
+// endpointPortNumber returns the port to be used for this endpoint. It is zero
+// if the endpoint does not match the given service port.
+func endpointPortNumber(servicePort corev1.ServicePort, endpointPorts []corev1.EndpointPort) int32 {
+	// Is this reasonable to assume?
+	if len(endpointPorts) == 0 {
+		return servicePort.Port
+	}
+
+	for _, endpointPort := range endpointPorts {
+		// For matching endpoints, the port names must correspond, either by
+		// being empty or non-empty. Multi-port services mandate non-empty
+		// names and allow us to filter for the right addresses.
+		if servicePort.Name == endpointPort.Name {
+			return endpointPort.Port
 		}
-		return int(port.Port)
 	}
-	return int(servicePort.Port)
+	return 0
```

Comment:
Functional, but man, does it suck debugging what `0` means. Can we provide/log a meaningful error?

Reply:
As [explained above](https://github.com/containous/traefik/pull/3121#discussion_r179962156), it's not an error but completely legitimate if we can't find a match. For such cases, it's common to return a special value denoting a "not found" state. For instance, [`strings.Index`](https://golang.org/pkg/strings/#Index) from the stdlib does the same.

We could also use the "comma ok" idiom (i.e., return an integer and a boolean where the latter indicates if a value was found). Looking at the stdlib again, however, that's only really used when a special value cannot be employed (such as for map lookups).

Personally, I find the chosen approach simple and idiomatic. I also made sure to document the return value semantics in the function's GoDoc.